src/full/Agda/Interaction/Highlighting/LaTeX/Base.hs

```diff
@@ -0,0 +1,807 @@
+import Data.Maybe
+import Data.Function
+import Data.Foldable (toList)
+import Control.Monad.Trans.Reader as R ( ReaderT(runReaderT) )
+import Control.Monad.RWS.Strict
+import Control.Arrow (second)
+import System.Directory
+import System.Exit
+import System.FilePath
+import System.Process
+import Data.Text (Text)
+import qualified Data.Text               as T
+#ifdef COUNT_CLUSTERS
+import qualified Data.Text.ICU           as ICU
+#endif
+import qualified Data.Text.Lazy          as L
+import qualified Data.Text.Lazy.Encoding as E
+import qualified Data.ByteString.Lazy    as BS
+
+import Data.HashSet (HashSet)
+import qualified Data.HashSet as Set
+import qualified Data.IntMap  as IntMap
+import qualified Data.List    as List
+
+import Paths_Agda
+
+import Agda.Syntax.Abstract (toTopLevelModuleName)
+import Agda.Syntax.Common
+import Agda.Syntax.Concrete (TopLevelModuleName, moduleNameParts)
+import Agda.Syntax.Parser.Literate (literateTeX, LayerRole, atomizeLayers)
+import qualified Agda.Syntax.Parser.Literate as L
+import Agda.Syntax.Position (startPos)
+
+import Agda.Interaction.Highlighting.Precise
+
+import Agda.TypeChecking.Monad (Interface(..))
+
+import Agda.Utils.FileName (AbsolutePath)
+import qualified Agda.Utils.List1 as List1
+
+import Agda.Utils.Impossible
+
+------------------------------------------------------------------------
+-- * Logging
+
+class Monad m => MonadLogLaTeX m where
+  logLaTeX :: LogMessage -> m ()
+
+-- | Log LaTeX messages using a provided action.
+--
+-- This could be accomplished by putting logs into the RWST output and splitting it
+-- into a WriterT, but that becomes slightly more complicated to reason about in
+-- the presence of IO exceptions.
+--
+-- We want the logging to be reasonbly polymorphic, avoid space leaks that can occur
+-- with WriterT, and also be usable during outer phases such as directory preparation.
+--
+-- I'm not certain this is the best way to do it, but whatever.
+type LogLaTeXT m = ReaderT (LogMessage -> m ()) m
+
+instance Monad m => MonadLogLaTeX (LogLaTeXT m) where
+  logLaTeX message = do
+    doLog <- ask
+    lift $ doLog message
+
+runLogLaTeXTWith :: Monad m => (LogMessage -> m ()) -> LogLaTeXT m a -> m a
+runLogLaTeXTWith = flip runReaderT
+
+-- Not currently used, but for example:
+-- runLogLaTeXWithIO :: MonadIO m => LogLaTeXT m a -> m a
+-- runLogLaTeXWithIO = runLogLaTeXTWith $ liftIO . T.putStrLn . logMsgToText
+
+------------------------------------------------------------------------
+-- * Datatypes.
+
+-- | The @LaTeX@ monad is a combination of @ExceptT@, @RWST@ and
+-- a logger @m@. The error part is just used to keep track whether we finished or
+-- not, the reader part contains static options used, the writer is where the
+-- output goes and the state is for keeping track of the tokens and some
+-- other useful info, and the MonadLogLaTeX part is used for printing debugging info.
+
+type LaTeX a = forall m. MonadLogLaTeX m => RWST Env [Output] State m a
```

Comment:
The comment here mentions `ExceptT` that does not occur in the definition of the `LaTeX` monad.

Also, the mixture of a `MonadLogLaTeX` constraint on the one hand and a concrete `RWST` implementation of the rest looks weird.  Why not use the `LogLaTeXT` transformer directly on top (or bottom) of `RWST` ?
If one wants the logging in `IO` one could run the `LogLaTeXT` with the `T.putStrLn . logMsgToText` implementation of the logging function.

Reply:
Turns out that `LogLaTeXT` is superfluous since we require `MonadLogLaTeX` from the base monad.
It can be deleted, and one can make a `newtype` wrapper for the base monad to implement `logLaTeX` in the desired way.  There is no need to go via `ReaderT`.